File: src/zjhmale/cps/CPSFoldingBuilder.java
package zjhmale.cps;

import com.intellij.lang.ASTNode;
import com.intellij.lang.folding.FoldingBuilder;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.util.TextRange;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import zjhmale.cps.setting.CPSSettings;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by zjh on 2016/2/19.
 */
public class CPSFoldingBuilder implements FoldingBuilder {
    private static final Pattern symbolPattern = Pattern.compile("\\(fn|\\(partial|\\(->|\\(def|not=|#\\(|#\\{|union|difference|intersection");
    private static final HashMap<String, String> prettySymbolMaps;
    private static final List<String> openDelimiters = Arrays.asList("(", "{", "[");
    private static final List<String> closeDelimiters = Arrays.asList(")", "}", "]");
    private static final List<String> setOperators = Arrays.asList("union", "difference", "intersection");

    static {
        prettySymbolMaps = new HashMap<String, String>();
        prettySymbolMaps.put("(fn", "λ");
        prettySymbolMaps.put("(partial", "Ƥ");
        prettySymbolMaps.put("(def", "≡");
        prettySymbolMaps.put("(defn", "ƒ");
        prettySymbolMaps.put("(->", "→");
        prettySymbolMaps.put("(->>", "⇉");
        prettySymbolMaps.put("not=", "≠");
        prettySymbolMaps.put("#(", "λ(");
        prettySymbolMaps.put("#{", "∈{");
        prettySymbolMaps.put("#{}", "∅");
        prettySymbolMaps.put("union", "⋃");
        prettySymbolMaps.put("intersection", "⋂");
        prettySymbolMaps.put("difference", "−");
    }

    private static boolean isContainOpenDelimiter(String text, int start) {
        boolean contain = false;
        String nextChar = "";
        while (!nextChar.equals("\n") && start < text.length()) {
            nextChar = text.substring(start, start + 1);
            if (openDelimiters.contains(nextChar)) {
                contain = true;
            }
            start++;
        }
        return contain;
    }

    private static boolean isDelimiterMatch(String text, int start) {
        String nextChar = "";
        int leftCount = 0;
        int rightCount = 0;
        while (!nextChar.equals("\n") && start < text.length()) {
            nextChar = text.substring(start, start + 1);
            if (openDelimiters.contains(nextChar)) {
                leftCount++;
            }
            if (closeDelimiters.contains(nextChar)) {
                rightCount++;
            }
            start++;
        }
        return rightCount > leftCount;
    }

    @NotNull
    @Override
    public FoldingDescriptor[] buildFoldRegions(@NotNull final ASTNode node, @NotNull final Document document) {
        CPSSettings settings = CPSSettings.getInstance();
        List<FoldingDescriptor> descriptors = new ArrayList<FoldingDescriptor>();
        final String text = node.getText();
        final Matcher matcher = symbolPattern.matcher(text);

        while (matcher.find()) {
            String key = text.substring(matcher.start(), matcher.end());
            final TextRange nodeRange = node.getTextRange();
            int rangeStart = nodeRange.getStartOffset() + matcher.start();
            int rangeEnd = nodeRange.getStartOffset() + matcher.end();
            boolean shouldFold = false;
            if (key.startsWith("(")) {
                rangeStart += 1;
            }

            String nextChar = text.substring(rangeEnd, rangeEnd + 1);

            if (key.equals("(def")) {
                if (nextChar.equals(" ")) {
                    if (isContainOpenDelimiter(text, rangeStart)) {
                        shouldFold = settings.turnOnDef && isDelimiterMatch(text, rangeStart);
                    }
                }
                if (!nextChar.equals(" ")) {
                    shouldFold = false;
                }
                if (nextChar.equals("n")) {
                    key = "(defn";
                    rangeEnd += 1;
                    shouldFold = settings.turnOnDefn;
                }
            } else if (key.equals("(fn")) {
                shouldFold = settings.turnOnFn;
            } else if (key.equals("(partial")) {
                shouldFold = settings.turnOnPartial;
            } else if (key.equals("(->")) {
                if (nextChar.equals(">")) {
                    key = "(->>";
                    rangeEnd += 1;
                    shouldFold = settings.turnOnThreadLast;
                }
                if (nextChar.equals(" ")) {
                    shouldFold = settings.turnOnThreadFirst && isDelimiterMatch(text, rangeStart);
                }
            } else if (key.equals("not=")) {
                shouldFold = settings.turnOnNotEqual && isDelimiterMatch(text, rangeStart);
            } else if (key.equals("#(")) {
                shouldFold = settings.turnOnLambda;
            } else if (key.equals("#{")) {
                if (nextChar.equals("}")) {
                    key = "#{}";
                    rangeEnd += 1;
                    shouldFold = settings.turnOnEmptySet;
                } else {
                    shouldFold = settings.turnOnSet;
                }
            } else if (setOperators.contains(key)) {
                String nextThreeChars = text.substring(rangeEnd, rangeEnd + 3);
                if (nextThreeChars.equals(" #{")) {
                    if (key.equals("union")) {
                        shouldFold = settings.turnOnSetUnion;
                    }
                    if (key.equals("difference")) {
                        shouldFold = settings.turnOnSetDifference;
                    }
                    if (key.equals("intersection")) {
                        shouldFold = settings.turnOnSetIntersection;
                    }
                }
            }

            if (shouldFold) {
                String pretty = prettySymbolMaps.get(key);
                final TextRange range = TextRange.create(rangeStart, rangeEnd);
                descriptors.add(new CPSFoldingDescriptor(node, range, null, pretty, true));
            }
        }
        return descriptors.toArray(new FoldingDescriptor[descriptors.size()]);
    }

    @Nullable
    @Override
    public String getPlaceholderText(@NotNull ASTNode node) {
        return null;
    }

    @Override
    public boolean isCollapsedByDefault(@NotNull ASTNode node) {
        return true;
    }
}

File: src/main/kotlin/zjhmale/cps/CPSFoldingBuilder.kt
package zjhmale.cps

import com.intellij.lang.ASTNode
import com.intellij.lang.folding.FoldingBuilder
import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.util.TextRange
import zjhmale.cps.setting.CPSSettings
import java.util.*
import java.util.regex.Pattern

/**
 * Created by zjh on 16/3/22.
 */
class CPSFoldingBuilder : FoldingBuilder {
    private val symbolPattern = Pattern.compile(
            "\\(fn|\\(let|\\(partial|\\(->|\\(def|\\(doseq|\\(comp|not=|and|or|not|>=|<=|#\\(|#\\{|union|difference|intersection"
    )

    private val prettySymbolMaps = hashMapOf(
            "(fn" to "λ",
            "(let" to "┝",
            "(letfn" to "λ",
            "(partial" to "Ƥ",
            "(def" to "≡",
            "(defn" to "ƒ",
            "(doseq" to "",
            "(comp" to "∘",
            "(->" to "→",
            "(->>" to "⇉",
            "not=" to "≠",
            "and" to "∧",
            "or" to "∨",
            "not" to "¬",
            ">=" to "≥",
            "<=" to "≤",
            "#(" to "λ(",
            "#{" to "∈{",
            "#{}" to "∅",
            "union" to "⋃",
            "intersection" to "⋂",
            "difference" to "−"
    )
    private val openDelimiters = listOf("(", "{", "[")
    private val closeDelimiters = listOf(")", "}", "]")
    private val setOperators = listOf("union", "difference", "intersection")
    private val leftStopFlags = listOf("(", "[", " ", "/")

    private val GT = { a: Int, b: Int -> a > b }
    private val GE = { a: Int, b: Int -> a >= b }

    private fun isDelimiterMatch(text: String, start: Int, op: (Int, Int) -> Boolean): Boolean {
        var startOffset = start
        var nextChar = ""
        var leftCount = 0
        var rightCount = 0
        while (nextChar != "\n" && startOffset < text.length) {
            nextChar = text.substring(startOffset, startOffset + 1)
            if (openDelimiters.contains(nextChar)) {
                leftCount++
            }
            if (closeDelimiters.contains(nextChar)) {
                rightCount++
            }
            startOffset++
        }
        return op(rightCount, leftCount)
    }

    private fun findLeftStopPos(text: String, start: Int): Int {
        var startOffset = start
        var prevChar = ""
        while (!leftStopFlags.contains(prevChar) && startOffset > 0) {
            prevChar = text.substring(startOffset - 1, startOffset)
            startOffset--
        }
        if (leftStopFlags.contains(prevChar)) {
            return startOffset
        } else {
            return -1
        }
    }

    override fun buildFoldRegions(node: ASTNode, document: Document): Array<out FoldingDescriptor> {
        val settings = CPSSettings.getInstance()
        val descriptors = ArrayList<FoldingDescriptor>()
        val text = node.text
        val matcher = symbolPattern.matcher(text)

        while (matcher.find()) {
            var key = text.substring(matcher.start(), matcher.end())
            val nodeRange = node.textRange
            var rangeStart = nodeRange.startOffset + matcher.start()
            var rangeEnd = nodeRange.startOffset + matcher.end()
            if (key.startsWith("(")) {
                rangeStart += 1
            }

            val nextChar = text.substring(rangeEnd, rangeEnd + 1)
            val prevChar = text.substring(rangeStart - 1, rangeStart)

            val shouldFold =
                    if (key == "(def") {
                        if (nextChar == " ") {
                            settings.turnOnDef && isDelimiterMatch(text, rangeStart, GE)
                        } else if (nextChar == "n") {
                            key = "(defn"
                            rangeEnd += 1
                            settings.turnOnDefn
                        } else {
                            false
                        }

                    } else if (key == "(fn") {
                        settings.turnOnFn
                    } else if (key == "(partial") {
                        settings.turnOnPartial
                    } else if (key == "(->") {
                        if (nextChar == ">") {
                            key = "(->>"
                            rangeEnd += 1
                            settings.turnOnThreadLast && isDelimiterMatch(text, rangeStart, GT)
                        } else if (nextChar == " ") {
                            settings.turnOnThreadFirst && isDelimiterMatch(text, rangeStart, GT)
                        } else {
                            false
                        }
                    } else if (key == "(let") {
                        val nextTwoChars = text.substring(rangeEnd, rangeEnd + 2)
                        if (nextTwoChars == "fn") {
                            key = "(letfn"
                            rangeEnd += 2
                            settings.turnOnLetfn && isDelimiterMatch(text, rangeStart, GT)
                        } else if (nextChar == " ") {
                            settings.turnOnLet && isDelimiterMatch(text, rangeStart, GT)
                        } else {
                            false
                        }
                    } else if (key == "not=") {
                        settings.turnOnNotEqual && isDelimiterMatch(text, rangeStart, GT)
                    } else if (key == ">=") {
                        if (prevChar == ">") {
                            false
                        } else {
                            settings.turnOnGT && isDelimiterMatch(text, rangeStart, GT)
                        }
                    } else if (key == "<=") {
                        settings.turnOnLT && isDelimiterMatch(text, rangeStart, GT)
                    } else if (key == "#(") {
                        settings.turnOnLambda
                    } else if (key == "#{") {
                        if (nextChar == "}") {
                            key = "#{}"
                            rangeEnd += 1
                            settings.turnOnEmptySet
                        } else {
                            settings.turnOnSet
                        }
                    } else if (setOperators.contains(key)) {
                        val leftStopPos = findLeftStopPos(text, rangeStart)
                        if (leftStopPos != -1 && leftStopFlags.contains(prevChar)) {
                            rangeStart = leftStopPos + 1
                            if (nextChar == " " || nextChar == "]") {
                                if (key == "union") {
                                    settings.turnOnSetUnion
                                } else if (key == "difference") {
                                    settings.turnOnSetDifference
                                } else if (key == "intersection") {
                                    settings.turnOnSetIntersection
                                } else {
                                    false
                                }
                            } else {
                                false
                            }
                        } else {
                            false
                        }
                    } else {
                        false
                    }

            if (shouldFold) {
                val pretty = prettySymbolMaps[key] ?: return arrayOf<FoldingDescriptor>()
                val range = TextRange.create(rangeStart, rangeEnd)
                descriptors.add(CPSFoldingDescriptor(node, range, null, pretty, true))
            }
        }
        return descriptors.toArray<FoldingDescriptor>(arrayOfNulls<FoldingDescriptor>(descriptors.size))
    }

    override fun getPlaceholderText(node: ASTNode) = null

    override fun isCollapsedByDefault(node: ASTNode) = true
}